gossip/src/contact_info.rs

```diff
@@ -0,0 +1,295 @@
+#![allow(dead_code)]
+#![allow(unused_variables)]
+#![allow(unused_imports)]
+pub use crate::legacy_contact_info::LegacyContactInfo;
+use {
+    crate::crds_value::MAX_WALLCLOCK,
+    serde::{Deserialize, Deserializer, Serialize, Serializer},
+    solana_sdk::{
+        pubkey::Pubkey,
+        sanitize::{Sanitize, SanitizeError},
+        serde_varint, short_vec,
+    },
+    static_assertions::const_assert_eq,
+    std::{
+        borrow::Cow,
+        collections::HashMap,
+        net::{IpAddr, SocketAddr},
+        num::TryFromIntError,
+    },
+    thiserror::Error,
+};
+
+const SOCKET_TAG_GOSSIP: u8 = 0;
+const SOCKET_TAG_REPAIR: u8 = 1;
+const SOCKET_TAG_RPC: u8 = 2;
+const SOCKET_TAG_RPC_PUBSUB: u8 = 3;
+const SOCKET_TAG_SERVE_REPAIR: u8 = 4;
+const SOCKET_TAG_TPU: u8 = 5;
+const SOCKET_TAG_TPU_QUIC: u8 = 6;
+const SOCKET_TAG_TPU_FORWARDS: u8 = 7;
+const SOCKET_TAG_TPU_FORWARDS_QUIC: u8 = 8;
+const SOCKET_TAG_TPU_VOTE: u8 = 9;
+const SOCKET_TAG_TVU: u8 = 10;
+const SOCKET_TAG_TVU_FORWARDS: u8 = 11;
+const_assert_eq!(SOCKET_CACHE_SIZE, 12);
+const SOCKET_CACHE_SIZE: usize = SOCKET_TAG_TVU_FORWARDS as usize + 1usize;
+
+#[derive(Debug, Error)]
+enum Error {
+    #[error("Invalid IP address index: {index}, num addrs: {num_addrs}")]
+    InvalidAddrIndex { index: u8, num_addrs: usize },
+    #[error("Invalid port: {0}")]
+    InvalidPort(/*port:*/ u16),
+    #[error("IP addresses saturated")]
+    IpAddrsSaturated,
+    #[error("Multicast IP address: {0}")]
+    MulticastAddr(IpAddr),
+    #[error("Socket tag not found: {0}")]
+    SocketNotFound(/*tag:*/ u8),
+    #[error("Unspecified IP address: {0}")]
+    UnspecifiedAddr(IpAddr),
+}
+
+#[derive(Clone, Debug, Default, Eq, PartialEq)]
+pub struct ContactInfo {
+    pubkey: Pubkey,
+    wallclock: u64,
+    shred_version: u16,
+    version: solana_version::Version,
+    addrs: Vec<IpAddr>,
+    // TODO: use port offset with varint? should be sorted!
+    sockets: Vec<(/*tag:*/ u8, /*addr-idx:*/ u8, /*port:*/ u16)>,
```

Comment:
how do we anticipate populating this `Vec`? we could potentially further compress by including `gossip_ip`, `gossip_port` and any other sockets that _DO NOT_ match `gossip_ip:(gossip_port+socket_tag)`.  that is, assume the common case of default cli args where all sockets are on the same IP and a consecutive range of port numbers starting from the gossip port

Reply:
You still need a bit which would say if this socket is defined off `gossip` socket or has a different `IpAddr`. So unless we do some convoluted bit packing it would be the same number of bytes.